File: interface/plant/py/hardware/plant_with_swing_up.py
# get quanser interface lib and tcp_protocol description (vscode[debuger] launched at root directory)
import sys
sys.path.append(r"C:\Quanser\0_libraries\python")
from pal.products.qube import QubeServo2, QubeServo3
from pal.utilities.math import SignalGenerator, ddt_filter
from pal.utilities.scope import Scope

sys.path.append(r"./communication/py")
import tcp_protocol_server as tcs

# init tcp host and port
HOST = '0.0.0.0'
PORT = 9999

# get other tools
from threading import Thread
import signal
import time
import math
import numpy as np

# Thread hanlder initialization
global KILL_THREAD
KILL_THREAD = False
def sig_handler(*args):
    global KILL_THREAD
    KILL_THREAD = True
signal.signal(signal.SIGINT, sig_handler)

# simulation time and plotting set
simulationTime = 30 # will run for 30 seconds
color = np.array([0, 1, 0], dtype=np.float64)

scopePendulum = Scope(
    title='Pendulum encoder - alpha (rad)',
    timeWindow=10,
    xLabel='Time (s)',
    yLabel='Position (rad)')
scopePendulum.attachSignal(name='Pendulum - alpha (rad)',  width=1)

scopeBase = Scope(
    title='Base encoder - theta (rad)',
    timeWindow=10,
    xLabel='Time (s)',
    yLabel='Position (rad)')
scopeBase.attachSignal(name='Base - theta (rad)',  width=1)

scopeVoltage = Scope(
    title='Motor Voltage',
    timeWindow=10,
    xLabel='Time (s)',
    yLabel='Voltage (volts)')
scopeVoltage.attachSignal(name='Voltage',  width=1)

# control-system scenario
def control_loop():
    # interface setting #
    # ------------------------------------------------ #
    # qube version, using hardware, pendulum
    qubeversion = 3
    hardware = 1
    pendulum = 1

    # frequency of system holder and sampler
    frequency = 50 # hz

    # for scope sampling rate
    countMax = frequency / 50
    count = 0

    # class initialization
    QubeClass = QubeServo3

    # for state estimation
    state_theta_dot = np.array([0,0], dtype=np.float64)
    state_alpha_dot = np.array([0,0], dtype=np.float64)
    state_alpha_f_dot = np.array([0,0], dtype=np.float64)
    state_alpha_f_2dot = np.array([0,0], dtype=np.float64)

    # swing-up standing gate
    stand_run = False
    change_flag = False
    set_time = 0
    switching_time = 1

    # gain of full-state(use initial part of swing-up only 50hz based)
    K = np.array([-2.0, 28.0, -1.5, 2.5])

    # gain P of swing up and bumpping gain R
    P = 0.015
    R = 1

    # switching target angle
    angle = 5

    # describe #
    # ------------------------------------------------ #
    # instance of hardware model 
    with QubeClass(hardware=hardware, pendulum=pendulum, frequency=frequency) as myQube:
        # instance of tcp layer
        with tcs.tcp_server(HOST, PORT) as tcsp:
            startTime = 0
            timeStamp = 0
            def elapsed_time():
                return time.time() - startTime
            startTime = time.time()

            while timeStamp < simulationTime and not KILL_THREAD:
                if not stand_run:
                    # read sensor information
                    myQube.read_outputs()

                    # calc output
                    theta = myQube.motorPosition * -1
                    alpha_f =  myQube.pendulumPosition
                    alpha = np.mod(alpha_f, 2*np.pi) - np.pi
                    alpha_deg = abs(math.degrees(alpha))

                    # Calculate angular velocities with filter of 50 and 100 rad
                    theta_dot, state_theta_dot = ddt_filter(theta, state_theta_dot, 50, 1/frequency)
                    alpha_dot, state_alpha_dot = ddt_filter(alpha, state_alpha_dot, 100, 1/frequency)
                    alpha_f_dot, state_alpha_f_dot = ddt_filter(alpha_f, state_alpha_f_dot, 100, 1/frequency)
                    alpha_f_2dot, state_alpha_f_2dot = ddt_filter(alpha_f_dot, state_alpha_f_2dot, 100, 1/frequency)
                    states = np.array([theta, alpha, theta_dot, alpha_dot])

                    # base limitation edge dectection
                    barrier_edge = np.array([0.0, 0.0])
                    barrier_flag = False
                    barrier_edge[1] = barrier_edge[0]
                    barrier_edge[0] = math.degrees(theta)

                    # barrier of base set
                    barrier_edge = np.array([0.0, 0.0])
                    barrier_flag = False
                    barrier_limit = 5

                    if(alpha_deg > angle and (not change_flag)):
                        barrier_edge[1] = barrier_edge[0]
                        barrier_edge[0] = math.degrees(theta)

                        if(barrier_edge[0] >= barrier_limit and barrier_edge[1] < barrier_limit):
                            barrier_flag = False
                        elif(barrier_edge[0] < -barrier_limit and barrier_edge[1] >= -barrier_limit):
                            barrier_flag = True

                        if(alpha_f_2dot > 0):
                            if(barrier_flag):
                                voltage = R
                            else:
                                voltage = P * abs(alpha_deg)
                        else:
                            if(not barrier_flag):
                                voltage = -R
                            else:
                                voltage = -P * abs(alpha_deg)
                    else:
                        voltage = 1*np.dot(K, states)
        
                        if(not change_flag):
                            change_flag = True
                            set_time = timeStamp
                            stand_run = True
                    
                    # write commands
                    myQube.write_voltage(voltage)

                    print(f"control start: {stand_run}")
                else:
                    # running signal send for controller
                    tcsp.send("run")
    
                    # read sensor information
                    myQube.read_outputs()
    
                    # calc output
                    theta = myQube.motorPosition * -1
                    alpha_f =  myQube.pendulumPosition
                    alpha = np.mod(alpha_f, 2*np.pi) - np.pi
                    alpha_deg = alpha * 180 / np.pi

                    # Calculate angular velocities with filter of 50 and 100 rad
                    theta_dot, state_theta_dot = ddt_filter(theta, state_theta_dot, 50, 1/frequency)
                    alpha_dot, state_alpha_dot = ddt_filter(alpha, state_alpha_dot, 100, 1/frequency)
                    states = np.array([theta, alpha, theta_dot, alpha_dot])
    
                    # send plant output
                    tcsp.send(-theta)
                    tcsp.send(-alpha)
    
                    # get control input
                    _, u = tcsp.recv()
    
                    # swing up trasient responce cushioning
                    if(timeStamp - set_time < switching_time):
                        voltage = 1*np.dot(K, states)
                        print("control object: full-state on plant")
                    else:
                        # running range set
                        if abs(alpha_deg) < 25:
                            voltage = u
                        else:
                            voltage = 0
                        print("control object: outside controller")
                    
                    # write commands
                    myQube.write_voltage(voltage)

                # plot to scopes
                count += 1
                if count >= countMax:
                    scopePendulum.sample(timeStamp, [-alpha])
                    scopeBase.sample(timeStamp, [-theta])
                    scopeVoltage.sample(timeStamp,[voltage])
                    count = 0

                timeStamp = elapsed_time()

            tcsp.send("end")

def main():
    thread_cl = Thread(target=control_loop)
    thread_cl.start()

    while thread_cl.is_alive() and (not KILL_THREAD):

        # This must be called regularly or the scope windows will freeze
        # Must be called in the main thread.
        Scope.refreshAll()
        time.sleep(0.01)

    input('Press the enter key to exit.')

if __name__ == "__main__":
    main()

File: interface/plant/cpp/hardware/plant_with_swing_up.cpp
#define _USE_MATH_DEFINES
#include <iostream>
#include <cstdint>
#include <string>
#include <cmath>
#include <chrono>
#include "tcp_protocol_server_windows.h"
#include "hil.h"
#include "quanser_timer.h"
using namespace std;

const string host = "0.0.0.0";
const int port = 9999;

double ddt_filter(double, vector<double>&, double, double);
void swing_up(vector<double>&, double&);

int main()
{
    t_card board;
    t_error result;
    // if you want to use hardware change to 1 else 0 is virtual(QLab)
    bool hardware = 0;

    if (hardware)
    {
        result = hil_open("qube_servo3_usb", "0", &board);
        if (result < 0)
        {
            cout << "failure to connect hardware" << endl;
            return -1;
        }
    }
    else
    {
        result = hil_open("qube_servo3_usb", "0@tcpip://localhost:18923?nagle='off'", &board);
        if (result < 0)
        {
            cout << "failure to connect QLab" << endl;
            return -1;
        }
    }

    // simulation_time is total run time, sample_time is sample_time.
    int simulation_time = 30;
    double sample_time = 0.02;
    t_timeout interval;
    t_timeout timeout;
    timeout_get_timeout(&interval, sample_time);
    timeout_get_current_time(&timeout);

    // angle[0] = base, angle[1] = pendulum
    double angle[2] = { 0.0, 0.0 };
    double voltage = 0.0;
    int32_t encoder_counts[2];
    uint32_t encoder_channels[2] = { 0, 1 };
    uint32_t analog_channels[2] = { 0 };
    uint32_t digital_channels[1] = { 0 };
    t_boolean digital_values[1] = { 1 };
    hil_write_digital(board, digital_channels, 1, digital_values);

    // swing-up standing gate
    bool stand_run = false;
    bool trans = false;
    double er = 0.2;
    int ei = int(1.0 / sample_time);
    int count = 0;

    // calculation angle
    double theta = 0.0;
    double alpha = 0.0;
    double alpha_deg = 0.0;

    // TCP/IP ready
    tcp_server tcsp = tcp_server(host, port);

    // Just check loop/total time
    auto stc = chrono::high_resolution_clock::now();
    auto edc = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::nanoseconds>(edc - stc);
    double run_time = duration.count() / 1000000;
    double stack_time = 0.0;

    // For swing-up
    vector<double> state(4);
    vector<double> state_theta_dot = { 0.0, 0.0 };
    vector<double> state_alpha_dot = { 0.0, 0.0 };
    vector<double> K = { -0.87865021, 18.27186509, -0.60044619, 1.53003176 }; // this params have to already known from matlab or python. (based on 20ms)

    // control loop
    for (int64_t i = 0; i < (int64_t)((double)simulation_time / sample_time); i++)
    {
        stc = chrono::high_resolution_clock::now();

        // time sleep
        timeout_add(&timeout, &timeout, &interval);
        qtimer_sleep(&timeout);

        // read sensor
        hil_read_encoder(board, encoder_channels, 2, encoder_counts);
        angle[0] = encoder_counts[0] * (2.0 * M_PI / 2048.0);
        angle[1] = encoder_counts[1] * (2.0 * M_PI / 2048.0);

        // calc output
        theta = -angle[0];
        alpha = fmod(angle[1], 2 * M_PI);
        if (alpha < 0) alpha += 2 * M_PI;
        alpha = alpha - M_PI;
        alpha_deg = alpha * 180 / M_PI;

        state[0] = theta;
        state[1] = alpha;
        state[2] = ddt_filter(theta, state_theta_dot, 50, sample_time);
        state[3] = ddt_filter(alpha, state_alpha_dot, 100, sample_time);

        if (!stand_run && !trans)
        {
            // check output 
            cout << "pendulum angle: " << alpha << endl;
            cout << "base angle: " << theta << endl;

            swing_up(state, voltage);

            if (abs(alpha) < er)
            {
                cout << "ready" << endl;
                trans = true;
            }
        }
        else if (!stand_run && trans)
        {
            voltage = 0.0;
            for (int i = 0; i < 4; i++)
            {
                voltage += K[i] * state[i];
            }

            // saturation
            if (voltage > 15)
            {
                voltage = 15.0;
            }
            else if (voltage < -15)
            {
                voltage = -15.0;
            }
            if (abs(alpha_deg) > 20)
            {
                voltage = 0.0;
            }

            if (count > ei)
            {
                cout << "set" << endl;
                stand_run = true;
            }
            else
            {
                count += 1;
            }
        }
        else
        {
            tcsp.Send<string>("run");

            tcsp.Send<double>(-theta);
            tcsp.Send<double>(-alpha);

            voltage = tcsp.Recv<double>();

            cout << "---------------------------------------------" << endl;
            cout << "pendulum angle: " << alpha << endl;
            cout << "base angle: " << theta << endl;
            cout << "control input: " << voltage << endl;

            // saturation
            if (voltage > 15)
            {
                voltage = 15.0;
            }
            else if (voltage < -15)
            {
                voltage = -15.0;
            }
            if (abs(alpha_deg) > 20)
            {
                voltage = 0.0;
            }
        }

        // actuator write
        hil_write_analog(board, analog_channels, 1, &voltage);

        edc = chrono::high_resolution_clock::now();
        duration = chrono::duration_cast<chrono::nanoseconds>(edc - stc);
        run_time = duration.count() / 1000000;
        stack_time += run_time / 1000;
        cout << "iter: " << i << " | loop time: " << run_time << "ms | total time: " << stack_time << "s" << endl;
    }

    tcsp.Send<string>("end");

    // logic terminate
    voltage = 0.0;
    hil_write_analog(board, analog_channels, 1, &voltage);
    digital_values[0] = 0;
    hil_write_digital(board, digital_channels, 1, digital_values);
    if (board != NULL)
    {
        hil_close(board);
    }

    return 0;
}

double ddt_filter(double u, vector<double>& state, double A, double Ts)
{
    double y = 1.0 / (A * Ts + 2.0) * (2.0 * A * u - 2 * A * state[0] - state[1] * (A * Ts - 2));

    state[0] = u;
    state[1] = y;
   
    return y;
}

void swing_up(vector<double>& state, double& u)
{
    double mp = 0.024;
    double Lp = 0.129;
    double l = Lp / 2;
    double g = 9.81;
    double Jp = mp * pow(Lp, 2) / 3;

    double E_ref = 0.0;
    double mu = 150.0;

    double E_pot = mp * g * l * (cos(state[1]) - 1);
    double E_kin = 0.5 * Jp * pow(state[3], 2);
    double E_total = E_pot + E_kin;
    double E_err = E_ref - E_total;
    
    double term = signbit(state[3] * cos(state[1])) ? -1.0 : 1.0;
    if (abs(state[3]) < 0.05 && abs(cos(state[1])) < 0.1)
    {
        term = 0.0;
    }
    u = -1 * mu * E_err * term;

    if (-state[0] > 0)
    {
        if (u > 0)
        {
            u = 0.0;
        }
    }
    else
    {
        if (u < 0)
        {
            u = 0.0;
        }
    }

    if (u > 8)
    {
        u = 8.0;
    }
    else if (u < -8)
    {
        u = -8.0;
    }
}